### ISSUE TITLE
Yolo label segment resampling breaks labels with holes in them

### ISSUE THREAD

**Issue by TheOfficialOzone** [ultralytics/ultralytics#18131]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

# When `update_labels_info` resamples polygons with holes in them it creates errors in the mask generation

Problem code _(Within data/dataset.py)_
![image](https://github.com/user-attachments/assets/2d19ae4e-920f-4568-bbeb-7944a5e41777)

### Explanation
Yolo resamples points along a segment/polygon. Normally this is fine, however, if a polygon with holes uses more than that upper limit of points; then the generated mask will have issues as it doesn't have enough detail to properly cover the object. This leads to pixels outside of the source polygon to be included & points within the source polygon to be excluded.

### What the mask is supposed to look like
![image](https://github.com/user-attachments/assets/f872746e-bf54-40d6-b55e-9e4d9271eb04)

### What Yolo generates
![image](https://github.com/user-attachments/assets/313824ab-4688-4c0e-a1fd-7291f37a6c09)

### _Mostly_ Fixed mask by increasing the sample count
![image](https://github.com/user-attachments/assets/42d2d5cf-72bf-4ff6-ad27-2a54a5ee26c7)

### If I decrease the sample count
![image](https://github.com/user-attachments/assets/b0fcc3ac-50a2-44a7-80a5-d01d6e2ae736)

## Other things to note
cv2.fillPoly() which I am using to visualize the mask (& what Yolo uses to generate them) also renders / includes infinitely thin lines
- IMO this is unintended behaviour. This should be changed so as no to include them in the mask

### Environment

Ultralytics 8.3.37 🚀 Python-3.10.12 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24564MiB)
Setup complete ✅ (32 CPUs, 15.5 GB RAM, 24.0/251.0 GB disk)

OS                  Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 15.47 GB
Disk                24.0/251.0 GB
CPU                 13th Gen Intel Core(TM) i9-13900
CPU count           32
GPU                 NVIDIA GeForce RTX 4090, 24564MiB
GPU count           1
CUDA                12.1

numpy               ✅ 2.1.2>=1.23.0
numpy               ✅ 2.1.2<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 5.4.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1>=1.8.0
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.11>=2.0.0

## Instructions
1. Download the provided training script & dataset
2. Run `python3 test_yolo.py`
3. Look at the generated 'train_batch0.jpg' in the "runs/segment" folder for the issue

Currently a temporary fix is to increase the samples, however that is a very bandaid solution for a larger issue.
 Unfortunately the fix for this is not as easy as removing the resampling as the rest of the code base depends on "segments" in "Instances" to be a numpy array, which cannot have arrays of different sizes contained within it. You likely need to rewrite the "Instances" object to not use numpy arrays in order to properly fix the issue.

Here is the training script
[script.zip](https://github.com/user-attachments/files/18062149/script.zip)

Here is the dataset that causes the issue to appear (1 image, but I can get more if needed)
[example_dataset.zip](https://github.com/user-attachments/files/18062102/example_dataset.zip)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TheOfficialOzone, thank you for bringing this up and for the detailed bug report 🐛! We appreciate your effort in providing images, context, and a dataset example—it’s incredibly helpful for understanding the issue.

This is an automated response 🤖, but rest assured that an Ultralytics engineer will review your issue shortly to assist further.

In the meantime, here are a few steps you can take to ensure everything is ready for debugging:

1. **Minimum Reproducible Example (MRE)**  
We noted that you’ve already provided steps and context for reproduction with associated files. Thank you for this! Just make sure your example simplifies the problem as much as possible while still reproducing the issue in case others attempt to replicate it.

2. **Upgrade your Ultralytics package**  
To ensure that your issue has not already been resolved in the latest version, please upgrade to the most recent `ultralytics` release:

```bash
pip install -U ultralytics
```

3. **Verified Environments**  
YOLO models can be run on various verified environments. For quick tests, you can try the latest `ultralytics` package in any of the following environments with preinstalled dependencies:  
   - Notebooks: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  

   For production environments, refer to our guides on [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/).

4. **Community Support and Resources** 💬  
For real-time help or discussions:  
   - Join [Discord](https://discord.com/invite/ultralytics) 🎧  
   - Post questions or collaborate on [Discourse](https://community.ultralytics.com)  
   - Explore our [Subreddit](https://reddit.com/r/Ultralytics) to engage with the community  

5. **Docs and Guides** 📚  
Explore the [Ultralytics Docs](https://docs.ultralytics.com) for technical usage, training tips, and troubleshooting solutions. For segment-related issues, reviewing the relevant guides under the **train** tasks section may provide additional insights.

Finally, if you’d like to attempt addressing the issue directly, you’re welcome to submit a Pull Request (PR) with any potential fixes or enhancements for review!

We appreciate your patience as we investigate further 🚀

**Comment by Y-T-G** [comment]
You can try the latest version

**Comment by Y-T-G** [comment]
There was [this fix](https://github.com/ultralytics/ultralytics/pull/18051) that was merged recently, which is why you should try the latest version.

**Comment by TheOfficialOzone** [comment]
So on the newest patch it is **better**, but it is not fixed for all cases

So it mostly fixes it for the example image I shared
![image](https://github.com/user-attachments/assets/f0a055a4-93dc-4044-a44b-32895b5db3c2)

However, in harder cases it still breaks
![image](https://github.com/user-attachments/assets/2ae156f0-801a-4eca-bce7-2d83fb007fc0)

I can still turn up the number of samples and it mostly fixes the issue in the harder case as well
![image](https://github.com/user-attachments/assets/b58e412d-c1a9-4e95-a820-1bfd94ddd5e9)

- There are still issues with the resampling, and I am curious as to why it's even necessary
- I also think that infinitely thin lines should not be included in the mask. So something other than cv2.fillpoly should be used, but I am not aware of any libraries that wouldn't include it in this case.
- Another fix would be a way to include "holes" as part of the mask, instead of using a trick of having infinitely thin lines between the holes in the mask


Here is the dataset of the two harder images
[harder_set.zip](https://github.com/user-attachments/files/18075642/harder_set.zip)

**Comment by Y-T-G** [comment]
> There are still issues with the resampling, and I am curious as to why it's even necessary

You wouldn't be able to create an array out of multiple segments or batch them for vectorized operations if they're of different sizes.

> Another fix would be a way to include "holes" as part of the labeling mask, instead of using a trick of having infinitely thin lines between the holes in the mask

Holes are currently not supported in the segmentation format because of how polygons work.

The type of segmentation you're trying to perform seems to be semantic segmentation, but YOLO is an instance segmentation model, which is why you have an issue.

You can try using [this](https://github.com/ultralytics/ultralytics/issues/17174#issuecomment-2439692715) but I'm not sure how well it works.

**Comment by TheOfficialOzone** [comment]
> You wouldn't be able to create an array out of multiple segments or batch them for vectorized operations if they're of different sizes.

For the resampling, would it be possible to have an option to keep the original bounds and just pad extra points into segments so that they match whichever segment is largest? That way it is still vectorizable and batchable. From what I understand so long as all segments within one Instance are the same size its all fine.
- For example: An image with two segments, one of length 20 & the other of length 5. The segment of length 5 would get padded to have a length of 20 by repeating the last point however many times is necessary.


> Holes are currently not supported in the segmentation format because of how polygons work.

I understand holes are not supported in the segmentation format and followed [this](https://github.com/ultralytics/ultralytics/issues/1106) example. This works perfectly when there a very few holes, but clearly in my case it starts to fails when the line to the start of the holes pass over an unincluded area.
- This is why I think cv2.fillPoly shouldn't be used to generate the masks. IMO Infinitely thin lines should not include any pixels, that is the behaviour I would expect.


> The type of segmentation you're trying to perform seems to be semantic segmentation, but YOLO is an instance segmentation model, which is why you have an issue.

Yes I am doing more of a semantic segmentation instead of instance segmentation (Specifically I am grouping clumps that are all connected as 1 detection instead of the 2 - 50 they otherwise should be). This is because creating training data for this is a herculean effort even when clumping them together (4ish hours per image for mediocre labels)

> You can try using https://github.com/ultralytics/ultralytics/issues/17174#issuecomment-2439692715 but I'm not sure how well it works.

I am actually already using a custom binary mask to yolo format as I was experiencing issues with the ones I had found online. It runs as a seperate preprocessing step and spits out the yolo labels in the previous datasets I shared from those binary masks.

**Comment by Y-T-G** [comment]
> For the resampling, would it be possible to have an option to keep the original bounds and just pad extra points into segments so that they match whichever segment is largest?

It might be possible. What value should it pad it with? Because 0.0 is valid coordinate. And negative values are out of bounds. It might also break the augmentation transforms and have certain edge cases.

> This is why I think cv2.fillPoly shouldn't be used to generate the masks. 

I am not sure what else you can use. I think this is a problem fundamentally with the format and not how it is being filled. It's hard to distinguish between points that are merely meant to connect different segments, and points that are normal.

**Comment by Y-T-G** [comment]
@Laughing-q Is it possible to pad segments instead of resampling?

**Comment by Laughing-q** [comment]
@TheOfficialOzone Hey thanks for reporting! I investigated this a bit and found it's actually because the original length of one of the segment labels is greater than 1000(the default resample number) so it's downgrading the labels and somehow breaks the original label. I've opened a PR to fix this and I tested within this PR even the harder sample you uploaded looks good to me.
main:
![pic-241211-1604-29](https://github.com/user-attachments/assets/acff3d43-e070-4463-be94-89c45cb60c0c)
within the PR:
![pic-241211-1604-16](https://github.com/user-attachments/assets/7a0292b5-8892-4b62-8a87-e9b3c1c26ac5)

FYI as for the thin lines I think it's the best trade-off we can take to save multiple segment parts in the polygon way, we need this thin lines to connect each parts.

**Comment by TheOfficialOzone** [comment]
> FYI as for the thin lines I think it's the best trade-off we can take to save multiple segment parts in the polygon way, we need this thin lines to connect each parts.

From what I understand from Graphics rasterization, points are only including in the polygon if the middle of the pixel is within the bounds of the polygon.

If you render a triangle with **PySimpleGUI** it won't include infinitely thin lines [(source code)](https://github.com/user-attachments/files/18092278/github_issue_test.zip)

I will double check this behaviour on a renderer using WGPU in like 8h to make sure this isn't just this random Python library's behaviour.

Screenshots (Look at the points I provide in graph.DrawPolygon)
![image](https://github.com/user-attachments/assets/d056c049-720a-4f87-ac40-285085176fd8)

![image](https://github.com/user-attachments/assets/583d17f3-1000-4d5f-b585-03e881d3366b)

**Comment by TheOfficialOzone** [comment]
> @Laughing-q Is it possible to pad segments instead of resampling?

So I actually think the existing resampling already "pads" the numbers properly. This is because regardless of the amount of samples, so long as the total number of samples is greater than that of the polygon you are sampling from; all of the source polygons vertices will be included in the resample. So the extra points generated from the resampling are always linearly interpolated from two source points forming a direct line which shouldn't effect the output.

Here is a test I wrote which performs the same operations as resample segments on a provided test array [source code](https://github.com/user-attachments/files/18092450/github_issue_test_2.zip)

I couldn't find a case where the source polygons points weren't all included in the final case. (Except when the samples were lower than the polygons total vertices!)

**Comment by Y-T-G** [comment]
I see. That makes sense. The previous PR I referred to would not be applied if the number of points were larger than n, so it wouldn't preserve the original points. So it fixed the issue for some of your images, and the others still had them. So @Laughing-q 's PR should fix that.

**Comment by TheOfficialOzone** [comment]
> I will double check this behaviour on a renderer using WGPU in like 8h to make sure this isn't just this random Python library's behaviour.

When using an actual graphics API it does not render infinitely thin lines. Within the next few days I will look into finding a python package that behaves this way & open a PR.

**Comment by glenn-jocher** [comment]
Thank you for investigating this further and for your commitment to improving the project! We look forward to your findings and any potential contributions via a PR. Your efforts are greatly appreciated.
